File: internal/maintenance/cat_plans.go
package maintenance

import (
	"context"
	"os"
	"path/filepath"
	"strings"
)

func init() { registerMatcher("plans", 60, scanPlans) }

// scanPlans lists every plan file under <root>/plans as a candidate — nothing
// is preselected, deletion is always an explicit choice. Staleness (older than
// the cutoff, default 60d) is surfaced as a Meta badge, NOT a candidacy filter:
// an old plan can be the only record of a design decision. Variant siblings (a
// plan plus its agent-variant files sharing a base name) are grouped so they're
// reviewed together; singletons stay ungrouped to avoid header noise.
func scanPlans(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	dir := filepath.Join(spec.Root, "plans")
	entries, ok, err := openDirNoSymlink(dir)
	if err != nil || !ok {
		return []Candidate{}, err
	}

	out := []Candidate{}
	for _, e := range entries {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		if e.IsDir() {
			continue
		}
		info, infoErr := e.Info()
		if infoErr != nil || info.Mode()&os.ModeSymlink != 0 {
			continue
		}
		mtime := info.ModTime()
		meta := map[string]string{"name": e.Name()}
		if olderThan(mtime, spec) {
			meta["stale"] = "true"
		}
		out = append(out, Candidate{
			Path: filepath.Join(dir, e.Name()), Bytes: info.Size(), Files: 1,
			ModTime: mtime, Reason: "plan document",
			Group: planBaseName(e.Name()), Meta: meta,
		})
	}

	// Ungroup singletons: a variant group only helps when siblings exist.
	counts := map[string]int{}
	for _, c := range out {
		counts[c.Group]++
	}
	for i := range out {
		if counts[out[i].Group] < 2 {
			out[i].Group = ""
		}
	}
	return out, nil
}

// planBaseName groups a plan with its variant siblings: the filename up to its
// first dot ("foo.md" and "foo.agent.md" → "foo").
func planBaseName(name string) string {
	if i := strings.IndexByte(name, '.'); i > 0 {
		return name[:i]
	}
	return name
}

File: src-tauri/src/maintenance/cat_plans.rs
//! Ported from `internal/maintenance/cat_plans.go` (W13). Lists every plan file
//! under `<root>/plans` as a candidate — nothing is preselected. Staleness (past
//! the cutoff, default 60d) is a Meta badge, NOT a candidacy filter. Variant
//! siblings (shared base name) are grouped; singletons stay ungrouped.

use std::collections::{BTreeMap, HashMap};
use std::path::Path;

use super::category::{mtime_utc, older_than, open_dir_no_symlink};
use super::types::{Candidate, CategorySpec};

pub fn scan_plans(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let dir = Path::new(&spec.root).join("plans");
    let (entries, ok) = open_dir_no_symlink(&dir)?;
    if !ok {
        return Ok(Vec::new());
    }

    let mut out: Vec<Candidate> = Vec::new();
    for e in &entries {
        if e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue;
        }
        let Ok(info) = e.metadata() else {
            continue;
        };
        if info.file_type().is_symlink() {
            continue;
        }
        let name = e.file_name().to_string_lossy().into_owned();
        let mtime = mtime_utc(&info);
        let mut meta = BTreeMap::new();
        meta.insert("name".to_string(), name.clone());
        if older_than(mtime, spec) {
            meta.insert("stale".to_string(), "true".to_string());
        }
        out.push(Candidate {
            path: dir.join(&name).to_string_lossy().into_owned(),
            bytes: info.len() as i64,
            files: 1,
            mod_time: mtime,
            reason: "plan document".to_string(),
            group: plan_base_name(&name).to_string(),
            meta,
        });
    }

    // Ungroup singletons: a variant group only helps when siblings exist.
    let mut counts: HashMap<String, usize> = HashMap::new();
    for c in &out {
        *counts.entry(c.group.clone()).or_insert(0) += 1;
    }
    for c in out.iter_mut() {
        if counts.get(&c.group).copied().unwrap_or(0) < 2 {
            c.group = String::new();
        }
    }
    Ok(out)
}

/// Groups a plan with its variant siblings: the filename up to its first dot
/// (index > 0). Mirrors Go `planBaseName`.
fn plan_base_name(name: &str) -> &str {
    match name.find('.') {
        Some(i) if i > 0 => &name[..i],
        _ => name,
    }
}

#[cfg(test)]
#[path = "cat_plans_tests.rs"]
mod cat_plans_tests;
